keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2397 @@
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Feature not used
+
+In a cluster where the feature is not used (no resource driver installed, no
+pods using dynamic resource allocation) the impact is minimal, both for
+performance and security. The scheduler plugin and resource controller will
+return quickly without doing any work for pods.
+
+#### Compromised node
+
+Kubelet is intentionally limited to read-only access for all new API types
+to prevent that a
+compromised kubelet interferes with scheduling of pending pods, for example
+by updating information normally published by the resource driver controller.
+Faking such information could be used for a denial-of-service
+attack against pods using those ResourceClaims, for example by overwriting
+their allocation result with a node selector that matches no node. A
+denial-of-service attack against the cluster and other pods is harder, but
+still possible. For example, frequently updating ResourceClaims could cause new
+scheduling attempts for pending pods. Filling up the `ReservedFor` field
+could exhaust etcd storage.
+
+Another potential attack goal is to get pods with sensitive workloads to run on
+a compromised node. For pods that don't use special resources nothing changes
+in that regard. Such an attack is possible for pods with extended resources
+because kubelet is in control of which capacity it reports for those: it could
+publish much higher values than the device plugin reported and thus attract
+pods to the node that normally would run elsewhere. With dynamic resource
+allocation, such an attack is still possible, but the attack code would have to
+be different for each resource driver because all of them will use their own,
+custom approach for reporting resource availability.
+
+The security of those custom approaches is the responsibility of the resource
+driver vendor. Solutions like Akri which establish their own control plane and
+then communicate with Kubernetes through the device plugin API already need to
+address this.
+
+#### User permissions and quotas
+
+Similar to generic ephemeral inline volumes, the [ephemeral resource use
+case](#ephemeral-vs-persistent-resourceclaims-lifecycle) gets covered by
+creating ResourceClaims on behalf of the user automatically through
+kube-controller-manager. The implication is that RBAC rules that are meant to
+prevent creating ResourceClaims for certain users can be circumvented, at least
+for ephemeral resources. Administrators need to be aware of this caveat when
+designing user restrictions.
+
+A quota system that limits how much of the underlying resources a user may consume
+needs to be supported by the resource driver. When a user has exhausted their
+quota, the driver then would refuse to allocate further ResourceClaims. Such a
+quota system cannot be implemented in core Kubernetes because Kubernetes has no
+information about how much a certain ResourceClaim would count against the quota.
```

Comment:
This seems really limiting. As a cluster admin, I can limit how many resource claims a namespace can make, but _not_ prevent a namespace from consuming more than it's fair share of the actual resource?

Isn't that liable to make the mechanism unusable for cluster operators who can't trust their (namespace-scoped) users?

Could we at least define a way to _record_ the limit, even if the enforcement mechanism is left up to the cluster operator?

Reply:
> As a cluster admin, I can limit how many resource claims a namespace can make, but not prevent a namespace from consuming more than it's fair share of the actual resource?

You can, assuming that the resource driver implements a quota system.

The point here is not that it cannot be done at all, just that it cannot be implemented *in core Kubernetes components*.

> Could we at least define a way to record the limit, 

A resource driver can define a CRD where the cluster operator configures such limits.

Reply:
Is there an opportunity to standardize this aspect? I suspect many vendors will end up implementing similar CRDs for limits and quotas.

Reply:
> async (claim just fails to resolve
> ...
> 2. Have driver-specific quota CRs and tell users to create an example.com/Quota rather than a standard one. 

This is what I meant with the text in the KEP. I didn't want to go into implementation details, but that could be changed by adding a non-binding "Implementing resource quota in a resource driver" section below.

Reply:
I think I'm fine to see this called out as a beta criterion - we don't need to know yet what the answer is, but we need an answer

Reply:
This is not an implementation details to me, instead it is more a end-to-end design flow for the feature. But I am ok we punt this to beta though.